protocol-select/README.md

```diff
@@ -0,0 +1,511 @@
+fails, they close the current connection and dial a new connection, this time
+using [Multistream Select].
+
+Note that it takes one connection attempt to discover this failure and an
+additional attempt to perform the fallback. Thus this upgrade mechanism should
+only be used in addition to the above multiphase rollout to ease the transition.
+
+<!-- TODO Find a better name than `Heuristics` -->
+### Heuristics
+
+When accepting a connection, an endpoint doesn't know whether the remote peer is
+going to speak [Multistream Select] or Protocol Select to negotiate connection
+or stream protocols. The below first describes **at which stage** of a TCP and
+QUIC connection the two protocol negotiation protocols need to be
+differentiated, followed by **how** one can differentiate the two.
+
+#### TCP
+
+The first message received on a freshly established and secured TCP connection
+will be a message trying to negotiate the stream muxer using either Protocol
+Select or [Multistream Select].
+
+#### QUIC
+
+Since QUIC provides native stream multiplexing, there's no need to negotiate
+a stream multiplexer. We therefore have to wait a bit longer before we can
+distinguish between [Multistream Select] and Protocol Select, namely until
+the dialer opens the first stream. Conversely, this means that a listener won't
+be able to open a stream until it has determined which protocol is used.
+
+#### Protocol Differentiation
+
+Both Protocol Select and [Multistream Select] prefix their messages with the
+varint encoded message length. The first message send by [Multistream Select] is
+`/multistream/1.0.0`. Implementations should read the first few bytes and
+proceed with either [Multistream Select] or Protocol Select depending on whether
+it equals `/multistream/1.0.0` or not.
+
+## Protocol Specification
+
+Messages are encoded according to the Protobuf definition below using the
+`proto2` syntax. The encoded messages are prefixed with their length in bytes,
+encoded as an unsigned variable length integer as defined by the [multiformats
+unsigned-varint spec][uvarint-spec].
+
+Messages are encoded via the `ProtoSelect` message type. With the current
+version of _Protocol Select_ detailed in this document, the `version` field of
+the `ProtocolSelect` message is set to `1`. Implementations MUST reject messages
+with a `version` other than the current version. See [Protocol
+Evolution](#Protocol-Evolution) for details. The `Protocol` `name` field is a
+UTF-8 encoded string identifying a specific protocol, same as in _Multistream
+Select_. See [Protocol Negotiation
+section](../connections/README.md#protocol-negotiation) for libp2p specific
+_Protocol Name_ conventions such as path-like structure.
+
+```protobuf
+syntax = "proto2";
+
+message ProtoSelect {
+    required uint32 version = 1;
+
+    message Protocol {
+        oneof protocol {
+            string name = 1;
+        }
+    }
+    repeated Protocol protocols = 2;
```

Comment:
:+1: This came up in previous discussions as well.

I would like to keep the first iteration of the protocol as simple as possible. As far as I can tell, this can be added in a backwards compatible way later on. Let me know in case I am missing something.

Reply:
Sure, though two things to keep in mind:

The "Optimistic Protocol Negotiation" from the current spec should probably be removed, to avoid clashes in the future

And to add backward compatibility in the future, we'll need need to add a "early data processed" bool to the listener -> initiator response, otherwise the initiator has no way to know if the early data has been processed. Not too much of a deal, though (I think)

Reply:
> And to add backward compatibility in the future, we'll need need to add a "early data processed" bool to the listener -> initiator response, otherwise the initiator has no way to know if the early data has been processed. Not too much of a deal, though (I think)

That would require the listener to only send a Protocol Select message once it received the initiators message, right? If so, that would prevent all early-data optimizations where the listener side is able to send first.

Reply:
Right, might be an issue.

In the case where the listener side is sending first, we would have to flip the roles (the listener is becoming initiator)
AFAICT, we always know who should be the "protocol select initiator", even though it may not be the same as the "secure initiator"

But maybe there is a case where we don't know?